fendermint/vm/topdown/src/sync.rs

```diff
@@ -3,139 +3,245 @@
 //! A constant running process that fetch or listener to parent state
 
 use crate::error::Error;
-use crate::{BlockHash, BlockHeight, Config, InMemoryFinalityProvider, ParentFinalityProvider};
+use crate::{
+    BlockHash, BlockHeight, CachedFinalityProvider, Config, IPCParentFinality,
+    ParentFinalityProvider, Toggle,
+};
 use anyhow::{anyhow, Context};
-use async_stm::atomically_or_err;
+use async_stm::{atomically, atomically_or_err};
 use fvm_shared::clock::ChainEpoch;
 use ipc_agent_sdk::apis::IpcAgentClient;
 use ipc_agent_sdk::jsonrpc::JsonRpcClientImpl;
 use ipc_agent_sdk::message::ipc::ValidatorSet;
 use ipc_sdk::cross::CrossMsg;
 use ipc_sdk::subnet_id::SubnetID;
+use std::cmp::min;
 use std::sync::Arc;
 use std::time::Duration;
 
+/// The max number of blocks polling should query each parent view update. If the number of blocks
+/// polled equals this value, it would stop polling for this iteration and commit the result to cache.
+const MAX_PARENT_VIEW_BLOCK_GAP: BlockHeight = 100;
+/// When polling parent view, if the number of top down messages exceeds this limit,
+/// the polling will stop for this iteration and commit the result to cache.
+const TOPDOWN_MSG_LEN_THRESHOLD: usize = 500;
+
+/// Query the parent finality from the block chain state
+pub trait ParentFinalityStateQuery {
+    /// Get the latest committed finality from the state
+    fn get_latest_committed_finality(&self) -> anyhow::Result<Option<IPCParentFinality>>;
+}
+
 /// Constantly syncing with parent through polling
-pub struct PollingParentSyncer {
+pub struct PollingParentSyncer<T> {
+    subnet_id: SubnetID,
     config: Config,
-    parent_view_provider: Arc<InMemoryFinalityProvider>,
+    parent_view_provider: Arc<Toggle<CachedFinalityProvider>>,
     agent: Arc<IPCAgentProxy>,
+    committed_state_query: Arc<T>,
 }
 
-impl PollingParentSyncer {
+/// Queries the starting finality for polling. First checks the committed finality, if none, that
+/// means the chain has just started, then query from the parent to get the genesis epoch.
+async fn query_starting_finality<T: ParentFinalityStateQuery + Send + Sync + 'static>(
+    subnet_id: &SubnetID,
+    query: &Arc<T>,
+    agent: &Arc<IPCAgentProxy>,
+) -> anyhow::Result<IPCParentFinality> {
+    loop {
+        let mut finality = match query.get_latest_committed_finality() {
+            Ok(Some(finality)) => finality,
+            Ok(None) => {
+                tracing::debug!("app not ready for query yet");
+                tokio::time::sleep(Duration::from_secs(5)).await;
+                continue;
+            }
+            Err(e) => {
+                tracing::warn!("cannot get committed finality: {e}");
+                tokio::time::sleep(Duration::from_secs(5)).await;
```

Comment:
Okay I see that there are no panics and no errors returned from the provider, so no race condition :+1: